src/implementations/vanilla/components/carousel/carousel.js

```diff
@@ -0,0 +1,388 @@
+      navigationItemsClass = '.ecl-carousel__navigation-item',
+      attachClickListener = true,
+      attachResizeListener = true,
+    } = {}
+  ) {
+    // Check element
+    if (!element || element.nodeType !== Node.ELEMENT_NODE) {
+      throw new TypeError(
+        'DOM element should be given to initialize this widget.'
+      );
+    }
+
+    this.element = element;
+
+    // Options
+    this.toggleSelector = toggleSelector;
+    this.prevSelector = prevSelector;
+    this.nextSelector = nextSelector;
+    this.containerClass = containerClass;
+    this.slidesClass = slidesClass;
+    this.slideClass = slideClass;
+    this.currentSlideClass = currentSlideClass;
+    this.navigationItemsClass = navigationItemsClass;
+    this.attachClickListener = attachClickListener;
+    this.attachResizeListener = attachResizeListener;
+
+    // Private variables
+    this.toggle = null;
+    this.container = null;
+    this.slides = null;
+    this.btnPrev = null;
+    this.btnNext = null;
+    this.index = 1;
+    this.total = 0;
+    this.allowShift = true;
+    this.autoPlay = false;
+    this.autoPlayInterval = null;
+    this.resizeTimer = null;
+    this.posX1 = 0;
+    this.posX2 = 0;
+    this.posInitial = 0;
+    this.posFinal = 0;
+    this.threshold = 80;
+    this.navigationItems = null;
+
+    // Bind `this` for use in callbacks
+    this.handleClickOnToggle = this.handleClickOnToggle.bind(this);
+    this.shiftSlide = this.shiftSlide.bind(this);
+    this.checkIndex = this.checkIndex.bind(this);
+    this.moveSlides = this.moveSlides.bind(this);
+    this.resizeTicker = this.resizeTicker.bind(this);
+    this.handleResize = this.handleResize.bind(this);
+    this.dragStart = this.dragStart.bind(this);
+    this.dragEnd = this.dragEnd.bind(this);
+    this.dragAction = this.dragAction.bind(this);
+  }
+
+  /**
+   * Initialise component.
+   */
+  init() {
+    this.toggle = queryOne(this.toggleSelector, this.element);
+    this.btnPrev = queryOne(this.prevSelector, this.element);
+    this.btnNext = queryOne(this.nextSelector, this.element);
+    this.slidesContainer = queryOne(this.slidesClass, this.element);
+    this.container = queryOne(this.containerClass, this.element);
+    this.navigationItems = queryAll(this.navigationItemsClass, this.element);
+    this.currentSlide = queryOne(this.currentSlideClass, this.element);
+
+    this.slides = queryAll(this.slideClass, this.element);
+    this.total = this.slides.length;
+
+    const firstSlide = this.slides[0];
+    const lastSlide = this.slides[this.slides.length - 1];
+    const cloneFirst = firstSlide.cloneNode(true);
+    const cloneLast = lastSlide.cloneNode(true);
+
+    // Clone first and last slide
+    this.slidesContainer.appendChild(cloneFirst);
+    this.slidesContainer.insertBefore(cloneLast, firstSlide);
+
+    // Refresh the slides variable after adding new cloned slides
+    this.slides = queryAll(this.slideClass, this.element);
```

Comment:
can't we just add the two new slides to the array, instead of doing another query?

Reply:
need to slide the nodeList to an array, then unshift and push, to me it's easier to keep nodeList by redo a query, no?